### PR TITLE
Integration tests for Infrastructure

### LINKED ISSUE #102: Add integration test for Infrastructure reconciliation and deletion
**How to categorize this issue?**

/area testing
/kind enhancement
/priority normal
/platform azure

**What would you like to be added**:
With https://github.com/gardener/gardener-extension-provider-aws/pull/54 provider-aws features integration test for Infrastructure reconciliation and deletion.

**Why is this needed**:
I would be good if we have similar test for provider-azure as it will increase the robustness.

### PR DESCRIPTION
**Pull request by kon-angelo**
**How to categorize this PR?**

/area quality
/kind test
/priority normal
/platform azure

**What this PR does / why we need it**:
Adds integration tests for the infrastructure controller

**Which issue(s) this PR fixes**:
Fixes #102 

**Special notes for your reviewer**:

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: test/integration/infrastructure/infrastructure_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package infrastructure_test
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_suite_test.go:1]: license header missing

File: test/integration/infrastructure/infrastructure_test.go [hunk 2]
```diff
@@ -0,0 +1,746 @@
+package infrastructure_test
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:1]: license header missing

File: test/integration/infrastructure/matchers.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
+package infrastructure
```
Comment (rfranzke) [inline test/integration/infrastructure/matchers.go:1]: license header missing

File: test/integration/infrastructure/matchers.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	"fmt"
+	"net/http"
+	"reflect"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/onsi/gomega/format"
+	"github.com/onsi/gomega/types"
+)
+
+type AzureNotFoundErrorMatcher struct{}
```
Comment (rfranzke) [inline test/integration/infrastructure/matchers.go:27]: What about some doc strings for these exported structs, functions, etc?

File: test/integration/infrastructure/infrastructure_test.go [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+		return nil, err
+	}
+	spToken, err := adal.NewServicePrincipalToken(*oauthConfig, clientId, clientSecret, azureRest.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	return autorest.NewBearerAuthorizer(spToken), nil
+}
+
+func newProviderConfig(vnet *azurev1alpha1.VNet, natGateway, zoned bool) *azurev1alpha1.InfrastructureConfig {
```
Comment (dkistner) [inline test/integration/infrastructure/infrastructure_test.go:433]: Can we rename that to `newInfrastructureConfig`?

File: test/integration/infrastructure/infrastructure_test.go [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+
+			namespace, err := generateName()
+			Expect(err).NotTo(HaveOccurred())
+
+			err = runTest(ctx, logger, c, clientSet, namespace, providerConfig, decoder)
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+
+	Context("with zonal cluster using existing vNet and NatGateway", func() {
```
Comment (dkistner) [inline test/integration/infrastructure/infrastructure_test.go:259]: Should we add also a test for existing vnet + availabilityset based cluster?

Comment (kon-angelo) [reply test/integration/infrastructure/infrastructure_test.go:315]: Added one, with the addition of a "foreign" identity

**Comment by rfranzke** [comment]
`49m52s` - why does it take so long? 👀

File: test/integration/infrastructure/infrastructure_test.go [hunk 7]
```diff
@@ -0,0 +1,825 @@
...
+			framework.RunCleanupActions()
+		})
+
+		It("should successfully create and delete", func() {
+			foreignName, err := generateName()
+			Expect(err).ToNot(HaveOccurred())
+
+			foreignGroupName := foreignName
+			foreignVNetName := foreignName
+			foreignIdName := foreignName
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:326]: Minor comment: We usually use `var ()` blocks when we declare multiple variables in a row. No necessity to change it now, just wanted to make you aware :)

File: test/integration/infrastructure/infrastructure_test.go [hunk 8]
```diff
@@ -0,0 +1,825 @@
...
+		By("delete infrastructure")
+		Expect(client.IgnoreNotFound(c.Delete(ctx, infra))).To(Succeed())
+
+		By("wait until infrastructure is deleted")
+		err := common.WaitUntilExtensionCRDeleted(
+			ctx,
+			c,
+			logger,
+			func() extensionsv1alpha1.Object { return &extensionsv1alpha1.Infrastructure{} },
+			"Infrastructure",
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:385]:
```suggestion
			extensionsv1alpha1.InfrastructureResource,
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 9]
```diff
@@ -0,0 +1,825 @@
...
+		return err
+	}
+	if err := c.Create(ctx, cluster); err != nil {
+		return err
+	}
+
+	By("deploy cloudprovider secret into namespace")
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cloudprovider",
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:424]:
```suggestion
			Name:      v1beta1constants.SecretNameCloudProvider,
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 10]
```diff
@@ -0,0 +1,825 @@
...
+		return err
+	}
+
+	By("wait until infrastructure is created")
+	if err := common.WaitUntilExtensionCRReady(
+		ctx,
+		c,
+		logger,
+		func() runtime.Object { return &extensionsv1alpha1.Infrastructure{} },
+		"Infrastucture",
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:454]:
```suggestion
		extensionsv1alpha1.InfrastructureResource,
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 11]
```diff
@@ -0,0 +1,825 @@
...
+	return nil
+}
+
+func generateName() (string, error) {
+	suffix, err := utils.GenerateRandomStringFromCharset(5, "0123456789abcdefghijklmnopqrstuvwxyz")
+	if err != nil {
+		return "", err
+	}
+
+	return "az-infrastructure-it--" + suffix, nil
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:487]:
```suggestion
	return "azure-infrastructure-it--" + suffix, nil
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 12]
```diff
@@ -0,0 +1,825 @@
...
+		},
+		Spec: extensionsv1alpha1.InfrastructureSpec{
+			DefaultSpec: extensionsv1alpha1.DefaultSpec{
+				Type: azure.Type,
+				ProviderConfig: &runtime.RawExtension{
+					Raw: providerConfigJSON,
+				},
+			},
+			SecretRef: corev1.SecretReference{
+				Name:      "cloudprovider",
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:595]:
```suggestion
				Name:      v1beta1constants.SecretNameCloudProvider,
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 13]
```diff
@@ -0,0 +1,825 @@
...
+func prepareNewResourceGroup(ctx context.Context, logger *logrus.Entry, az *azureClientSet, groupName, location string) error {
+	logger.Infof("generating new ResourceGroups: %s", groupName)
+	_, err := az.groups.CreateOrUpdate(ctx, groupName, resources.Group{
+		Location: pointer.StringPtr(location),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:613]:
```suggestion
	return err
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 14]
```diff
@@ -0,0 +1,825 @@
...
+	if err = vNetFuture.WaitForCompletionRef(ctx, az.vnet.Client); err != nil {
+		return err
+	}
+
+	_, err = vNetFuture.Result(az.vnet)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:641]:
```suggestion
	return err
```

**Comment by kon-angelo** [comment]
@rfranzke Maybe I am wrong, but it spent a lot of time in pending state.
 
+ first log: `2020-09-10 16:51:37 | directories:`
+ test start: `2020-09-10 17:26:07 | Will run 4 of 4 specs`
+ test end: 
```
2020-09-10 17:42:11 | SUCCESS! -- 4 Passed \| 0 Failed \| 0 Pending \| 0 Skipped
2020-09-10 17:42:11 | Ran 4 of 4 Specs in 964.542 seconds
```

File: test/integration/infrastructure/infrastructure_test.go [hunk 15]
```diff
@@ -0,0 +1,825 @@
...
+func prepareNewIdentity(ctx context.Context, logger *logrus.Entry, az *azureClientSet, groupName, idName, location string) error {
+	logger.Infof("generating new Identity %s/%s", groupName, idName)
+	_, err := az.msi.CreateOrUpdate(ctx, groupName, idName, msi.Identity{
+		Location: pointer.StringPtr(location),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (rfranzke) [inline test/integration/infrastructure/infrastructure_test.go:653]:
```suggestion
	return err
```

**Review by rfranzke — COMMENTED** [review]
Very nice PR, very good! Thanks a lot for your effort :) I mainly have nit comments

**Comment by rfranzke** [comment]
@kon-angelo OK, thanks. Maybe the test machinery cluster was full/busy at that point in time? Let's try it again:

/test-single

**Comment by schrodit** [comment]
The test itself took 16 minutes but we had some autoscaling issues with the testmachinery at that time.
The next tests should be faster

**Comment by rfranzke** [comment]
@schrodit thanks! While you're here - why doesn't `/test-single` alone work? Shouldn't it default to this test run?

/test-single .ci/testruns/default.yaml

**Comment by schrodit** [comment]
`test-single` does only work when the `tm-config.yaml` is configured in the main branch.

**Comment by rfranzke** [comment]
OK, thanks... :) I always forget these details. Will hopefully become better now that we'll make more use of the TestMachinery bot :)

**Comment by schrodit** [comment]
I'm also trying to improve some error messages and commands today

**Review by dkistner — APPROVED** [review]
@kon-angelo and I did a pair review together.

From my side it looks good. Thank you very much.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**